### PR TITLE
refactor(triggers): place triggers in classes

### PR DESCRIPTION
**Pull request by boc-tothefuture**
This refactors triggers into classes to prevent having so many methods pulled into the rule context.

### CODE REVIEW

**Comment by jimtng** [comment]
I'm looking into this but I'm a bit slow

**Review by jimtng — COMMENTED** [review]
Great PR! I only have a few suggestions/questions below. I am wondering if there's a way to DRY the range_trigger and proc_trigger methods, because they'd also be used in commands trigger, which I assume you're planning on doing in a separate PR.

File: lib/openhab/log/logger.rb [hunk 1]
```diff
@@ -61,6 +61,22 @@ def initialize(name)
         define_method("#{level}_enabled?") { @sl4fj_logger.send("is_#{level}_enabled") }
       end
 
+      #
+      # Logs a map of key(value) with an optional preamble at trace level
+      # @param [String] preamble to put at start of log message
+      # @param [Hash] key and values to log
+      def state(preamble = 'State:', **kwargs)
+        return unless trace_enabled?
```
[the remark this answers is not in the document]
Comment (boc-tothefuture) [reply lib/openhab/log/logger.rb:69]: It is dynamically defined above.

File: lib/openhab/dsl/rules/triggers/trigger.rb [hunk 2]
```diff
@@ -11,132 +9,66 @@ module Rules
...
-          # @return [String] A time of day trigger
-          TIME_OF_DAY = 'timer.TimeOfDayTrigger'
+          #
+          # Separates groups from items, and flattens any nested arrays of items
+          #
+          # @param [Array] item_array Array of items passed to a trigger
+          #
+          # @return [Array] A new flat array with any GroupMembers object left intact
+          #
+          def self.separate_groups(item_array)
```
[the remark this answers is not in the document]
Comment (boc-tothefuture) [reply lib/openhab/dsl/rules/triggers/trigger.rb:28]: Sure, I just moved it and didn't rename it.

**Comment by boc-tothefuture** [comment]
> Great PR! I only have a few suggestions/questions below. I am wondering if there's a way to DRY the range_trigger and proc_trigger methods, because they'd also be used in commands trigger, which I assume you're planning on doing in a separate PR.

I don't see a great way to DRY those, without making it more complicated.  In the end, range calls proc and proc needs to call either update or changed trigger and those will have different signatures.  We could pass in a block, but I think that would make it more complicated and harder to follow.

I am open to suggestions however for ways to do it.

**Comment by boc-tothefuture** [comment]
@jimtng did you have any suggestions for DRY here?  I don't see a simple way.

**Comment by jimtng** [comment]
> @jimtng did you have any suggestions for DRY here? I don't see a simple way.

I did take a closer look at it, and I couldn't come up with a good way to DRY it either. Happy to leave it as is.

**Pull request merged**